src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -445,72 +579,49 @@ public StructImpl download(
 
-    ImmutableList<URL> urls =
-        getUrls(
-            url,
-            /*ensureNonEmpty=*/ !allowFail,
-            /*checksumGiven=*/ !Strings.isNullOrEmpty(sha256) || !Strings.isNullOrEmpty(integrity));
-    Optional<Checksum> checksum;
-    RepositoryFunctionException checksumValidation = null;
-    try {
-      checksum = validateChecksum(sha256, integrity, urls);
-    } catch (RepositoryFunctionException e) {
-      checksum = Optional.<Checksum>empty();
-      checksumValidation = e;
+    if (defer) {
+      return request;
     }
 
-    StarlarkPath outputPath = getPath("download()", output);
-    WorkspaceRuleEvent w =
-        WorkspaceRuleEvent.newDownloadEvent(
-            urls,
-            output.toString(),
-            sha256,
-            integrity,
-            executable,
-            getIdentifyingStringForLogging(),
-            thread.getCallerLocation());
-    env.getListener().post(w);
-    Path downloadedPath;
-    try (SilentCloseable c =
-        Profiler.instance().profile("fetching: " + getIdentifyingStringForLogging())) {
-      checkInOutputDirectory("write", outputPath);
-      makeDirectories(outputPath.getPath());
-      downloadedPath =
-          downloadManager.download(
-              urls,
-              authHeaders,
-              checksum,
-              canonicalId,
-              Optional.<String>empty(),
-              outputPath.getPath(),
-              env.getListener(),
-              envVariables,
-              getIdentifyingStringForLogging());
-      if (executable) {
-        outputPath.getPath().setExecutable(true);
-      }
-    } catch (IOException e) {
-      if (allowFail) {
-        return StarlarkInfo.create(
-            StructProvider.STRUCT, ImmutableMap.of("success", false), Location.BUILTIN);
-      } else {
-        throw new RepositoryFunctionException(e, Transience.TRANSIENT);
-      }
-    } catch (InvalidPathException e) {
-      throw new RepositoryFunctionException(
-          Starlark.errorf("Could not create output path %s: %s", outputPath, e.getMessage()),
-          Transience.PERSISTENT);
+    PendingDownload download = startDownload(request, thread);
+    return completeDownload(download);
+  }
+
+  @StarlarkMethod(
+      name = "download_multiple",
+      useStarlarkThread = true,
+      parameters = {
+          @Param(
+              name = "requests",
+              allowedTypes = { @ParamType(type = Sequence.class, generic1 = Dict.class)},
+              doc = "TODO"
+          )
+      },
+      doc = "TODO"
+  )
+  public Sequence<StructImpl> downloadMultiple(Sequence<?> requestsUnchecked, StarlarkThread thread)
+      throws EvalException, RepositoryFunctionException, InterruptedException {
+    Sequence<DownloadRequest> requests =
+        Sequence.cast(requestsUnchecked, DownloadRequest.class, "requests");
+
+    List<PendingDownload> downloads = new ArrayList<>();
```

Comment:
could we merge `DownloadRequest` and `PendingDownload`? Then we only need a single class that essentially just wraps a `Future`. That does mean that the download starts as soon as you call `rctx.download(defer=True)`, but maybe that's a good thing.

Also at that point we could just let people wait on individual deferred downloads? That removes the need to return the awkward list of returned structs. `rctx.download(defer=True).execute()` can return what you expect.

Reply:
I thought the `rctx.download(defer=True).execute()` approach would make it impossible to do multiple downloads at the same time?

In the current implementation, the download doesn't actually start until the `download_multiple()` call, which is handy because it makes state management a bit easier because then it's certain that there are no HTTP downloads pending, except during the `download_multiple()` call. Do you think we should break that invariant and start the download immediately after the `download()` call and make things more complicated or do I misunderstand what you suggest above?

(then you'd be advocating for the `token` approach I mentioned here: https://github.com/bazelbuild/bazel/issues/19674#issuecomment-1753046186)

Reply:
My line of reasoning was that we should absolutely, positively, under no circumstances should leave a download running that was initiated by a repository rule if the processing of that repository rule terminates and this invariant seemed to be much easier to uphold if we also have the invariant "we only ever have downloads running when some sort of downloading function is on the call stack".

I'll take a look how dangerous it looks like to do it with the token approach as time allows (although now that Thanksgiving is over, my time is much more dear)

Reply:
Actually, it ended up being much more pleasant than I thought it would be.